### ISSUE TITLE
multilingual scaling laws predictions

### ISSUE THREAD

**Issue by dianaonutu** [OpenEuroLLM/Taskboard#302]
Fit "basic"/simple scaling laws on the multilingual models, as Niccolo did on the English models.

**Comment by dianaonutu** [comment]
- [ ] Expand the grid for smaller models on LUMI: 0.1B_ne extend to 3 additional HP points
- [ ] modeling (N,D, GBS, LR) multilingual, monolingual @aaronkl @dianaonutu 
- [ ] enrich the scaling law with predicted points from Stable-Decay analysis

**Comment by dianaonutu** [comment]
### Defining $N$ and $C$

- **How should we define $N$ so that the approximation for training compute $C$ is accurate?** 
- **Which $N$ definition should we use in 1) fitting the scaling and 2) making predictions?**
- **What should we account for when defining $N$ and $C$?**

#### Model parameters

Let $N$ denote the model size, $D$ the number of training tokens, $d$ the model dimension, and $v$ the vocabulary size.

A language model can be decomposed into three parameter groups: the input embeddings $N_\text{emb}=vd$, the transformer blocks $N_\text{tr}$, and the output projection (LM head / output embeddings) $N_\text{head}=vd$. Thus, for models with untied embeddings, the total number of trainable parameters is

$$
N_\text{tot} = N_\text{emb} + N_\text{tr} + N_\text{head}.
$$

When input and output embeddings are tied, the embedding matrix and LM head share parameters, giving

$$
N_\text{tot} = N_\text{emb} + N_\text{tr}.
$$

> **Note**: The term *non-embedding parameters* is ambiguous in the presence of tied embeddings. Excluding the embeddings yields

 >$$
 N_\text{tot} - N_\text{emb}
=
> \begin{cases}
> N_\text{tr} + N_\text{head}, & \text{untied embeddings}, \\
> N_\text{tr}, & \text{tied embeddings}.
> \end{cases}
> $$
>
> In the tied case, the LM head shares the same parameters as the input embeddings. Consequently, *non-embedding parameters* do not correspond to a unique quantity and depend on the embedding scheme.


#### Training compute

The training compute is commonly approximated as $C \approx 6ND$, where $C$ denotes the total training FLOPs. This approximation assumes that the number of FLOPs per token is proportional to the number of model parameters.

However, this relationship is only approximate, as different components of the model contribute differently to the total computational cost. In particular, the attention operations incur additional FLOPs that are not fully captured by parameter counts alone (Porian et al. (2024)). Consequently, several works introduce an attention-equivalent parameter count $N_\text{attn} = ndl$, where $n$ denotes the sequence length, and $l$ the number of transformer layers. Although $N_\text{attn}$ does not correspond to trainable parameters, it provides a measure of the additional computational cost associated with attention and is incorporated into effective parameter counts such as $N_\text{eff}$.

#### Literature Review
Below, we review these definitions using the parameter decomposition introduced above.

| Work | Definition of $N$ | Embeddings | LM Head | Tied |
|------|-------------------|------------|------|--------|
| Kaplan et al. (2020) | $N_{\text{Kaplan}} = N_\text{tr}$ | Excluded | Excluded | Yes <sub>Porian et al. (2024)</sub> |
| Porian et al. (2024) | $N_\text{Porian} = N_\text{tr} + N_\text{head}$ | Excluded | Included | No |
| Porian et al. (2024) - all non-embedding learnable parameters| $N_{\text{exact}} = N_\text{Porian} + \text{LayerNorm gains}$  | Excluded | Included| No
| Porian et al. (2024) - accounts for attn cost | $N_{\text{eff}}= N_\text{Porian} + N_\text{attn}$ | Excluded | Included | No |
| Hoffmann et al. (2022) Approach 1 & 2 | $N_{\text{eff}}$ <sub>Porian et al. (2024)</sub> | Excluded | Included | No |
| Hoffmann et al. (2022) Approach 3| $N_\text{Hoffmann} = N_\text{emb} + N_\text{tr} + N_\text{head}$ <sub>Porian et al. (2024)</sub> | Included | Included | No |
| Münninghoff et al. (2024) | $N_\text{Porian}$ <sub>Porian et al. (2024)</sub> | Excluded | Included | -|
| Gadre et al. (2024) | $N_\text{Porian}$ <sub>Porian et al. (2024)</sub> | Excluded | Included | -|
| Hu et al. (2024) | $N_\text{Porian}$ <sub>Porian et al. (2024)</sub> | Excluded | Included | -|
| DeepSeek (2024) | $N_{\text{Kaplan}}$ <sub>Porian et al. (2024)</sub>| Excluded | Included | - |
| DeepSeek (2024)| $N_\text{Porian}$ <sub>Porian et al. (2024)</sub> | Excluded | Included | - |
| DeepSeek (2024) - chosen one | hybrid of $N_{\text{Kaplan}}$ and  $N_\text{Porian}$ <sub>Porian et al. (2024)</sub>| Excluded | Included | - |

**Comment by dianaonutu** [comment]
## How does Megatron-LM calculate the number of parameters reported in the logs?

**Parameter counting in Megatron-LM**: all the logic lives in [theoretical_memory_usage.py:L12-186](https://github.com/OpenEuroLLM/NVIDIA-Megatron-LM/blob/055f7defc2284738b207c81c5cdc9f1fdc9dc897/megatron/training/theoretical_memory_usage.py#L12-L186).

### 1. "Number of parameters in transformer block" (line 122–126)

This is the sum of all transformer layer parameters plus the final layer norm, but excluding embeddings:
```
transformer_block = (dense_layers × params_per_dense_layer)
                  + (moe_layers   × params_per_moe_layer)
                  + final_layernorm  (= 2 × hidden_size)
```
Each dense layer includes attention projections + FFN (with SwiGLU multiplier) + 2 layer norms.

### 2. "Number of parameters in embedding layers" - tied vs. untied (lines 116–121)
```
embedding_size = hidden_size * padded_vocab_size

if args.untie_embeddings_and_output_weights:
    num_parameters_in_embedding_layers = 2 * embedding_size  # input + output embeddings
else:
    num_parameters_in_embedding_layers = embedding_size      # shared weights, counted once
```
So with **tied embeddings** (the default), only one copy of `hidden_size × vocab_size` is counted. With `--untie-embeddings-and-output-weights`, it doubles.

### 3. "Total number of parameters" (line 131–135)
```
total = transformer_block + mtp_block + embedding_layers
```

## Which numbers should we use from the Megatron logs to calculate $N_\text{Porian}$ for the scaling laws?
$N_\text{Porian}$ from [Porian et al. 2024](https://arxiv.org/abs/2406.19146)

### Case 1: Model with tied embeddings
 $N_\text{Porian} = \text{Total number of parameters}$ 

### Case 2: Model with untied embeddings
 $N_\text{Porian} = \text{Number of parameters in transformer block} + \text{Number of parameters in embedding layers} / 2$

**Comment by dianaonutu** [comment]
# Scaling Laws Loss $L(N_\text{Porian},D)$ Parametric Fit
> Note: The HP grid hasn't yet been entirely trained, thus some points might be suboptimal.

## 1. Fit on all collected data
**Separable** $L(N_\text{Porian}, D) = E + A N_\text{Porian}^{-\alpha} + B D^{-\beta}$

**Multiplicative** $L(N_\text{Porian}, D) = E + A N_\text{Porian}^{-\alpha} + B D^{-\beta} + G(ND)^{-\gamma}$

| Law | Models | $E$ | $A$ | $\alpha$ | $B$ | $\beta$ | $G$ | $\gamma$ |
|---|---|---|---|---|---|---|---|---|
| Separable | English | 1.5957 | 262.8 | 0.3162 | 216.4 | 0.2774 | – | – |
| Separable | Multilingual | 1.824 | 13819.7543 | 0.517 | 1140.8865 | 0.345 | – | – |
| Multiplicative | English | 1.3081 | 213.1 | 0.2837 | 181.5 | 0.2352 | −1556.7 | 0.1921 |
| Multiplicative | Multilingual | 709.27 | 329.91 | 0.0267 | 667.01 | 0.0095 | −1665.91 | 0.0034 |

<img width="4759" height="1588" alt="Image" src="https://github.com/user-attachments/assets/95325580-a8af-4a7f-9c15-129ef5d35b22" />

<img width="4759" height="1588" alt="Image" src="https://github.com/user-attachments/assets/cc57af7a-812e-4b3d-b21a-b285c44b8439" />

We observed similar behavior of residuals as in the monolingual scaling laws. Below, same plots, different x-axis:

<img width="4759" height="1587" alt="Image" src="https://github.com/user-attachments/assets/74622d85-ea06-43fe-a936-89c733758ef1" />

<img width="4759" height="1587" alt="Image" src="https://github.com/user-attachments/assets/70232d9f-b51b-49f8-b2e3-cf4cc30e5e78" />

## 2. Prediction at larger scales
Using scaling laws fit from above, training data up to $N_\text{Porian}$.
| $N_\text{Porian}$ | $D$ | Separable Law<br><sub>Mean (95% CI)</sub> | Interaction Law<br><sub>Mean (95% CI)</sub> | Empirical |
|------------------:|-----:|:-----------------------------------------:|:-------------------------------------------:|----------:|
| 1.9B | 300B | 2.166294<br><sub>[2.144802, 2.192945]</sub> | 2.141324<br><sub>[2.130338, 2.150412]</sub> | 2.144906 |
| 4.3B | 300B | 2.090031<br><sub>[2.050277, 2.132493]</sub> | 2.085098<br><sub>[2.043651, 2.114017]</sub> | 2.033639 |
| 8.0B | 300B | 2.049907<br><sub>[1.994854, 2.104353]</sub> | 2.087303<br><sub>[1.999807, 2.143593]</sub> | – |
| 8.0B | 1000B <br><sub>(Baby model)</sub> | 2.007552<br><sub>[1.946183, 2.069444]</sub> | 2.017932<br><sub>[1.932380, 2.075560]</sub> | – |


## 3. Prediction on held out points $N_\text{Porian} = 1.26B$:

<img width="2370" height="1485" alt="Image" src="https://github.com/user-attachments/assets/9cd58505-d182-4f59-a6b2-435d8da527fa" />

<img width="2370" height="1485" alt="Image" src="https://github.com/user-attachments/assets/7e8e930a-489c-4ffb-98f3-3cc8161d77fd" />

# Scaling Laws Loss $L(C)$

**Comment by JeniaJitsev** [comment]
What is still missing here are predictions on held out points. Fit up to cut off compute $C_{threshold}$, predict for held out points above, estimate the error and CI. See https://github.com/LAION-AI/scaling-laws-for-comparison/ and https://openreview.net/forum?id=cWnZLIdeKn.  Please add that part.

**Comment by JeniaJitsev** [comment]
Also downstream evals for each measured point are to be performed on full set of evals (at least DCLM core and math/code/reasoning), lets add this.

**Comment by dianaonutu** [comment]
Validation loss $C$, $N$, $D$

<img width="5830" height="1555" alt="Image" src="https://github.com/user-attachments/assets/91d8991e-7fbd-490d-9218-4cabfef8c201" />

**Comment by dianaonutu** [comment]
<img width="2314" height="1496" alt="Image" src="https://github.com/user-attachments/assets/cbfb4f9a-3b26-4a33-88ad-6a2a13922570" />

**Comment by dianaonutu** [comment]
<img width="2370" height="1485" alt="Image" src="https://github.com/user-attachments/assets/d140bfaf-139f-4330-9b9e-2377c3acb161" />

<img width="2370" height="1485" alt="Image" src="https://github.com/user-attachments/assets/fa07d42f-f71d-4c66-8626-aa17ae1ddece" />

**Comment by dianaonutu** [comment]
<img width="2370" height="1485" alt="Image" src="https://github.com/user-attachments/assets/3752a1e9-3652-4d34-82c2-c9db89d3eb8d" />
<img width="2370" height="1485" alt="Image" src="https://github.com/user-attachments/assets/4b0299e7-15af-4145-9a4a-44f90b9ac4f8" />

**Comment by dianaonutu** [comment]
LR prediction

| Model   | LR     | GBSZ     |
|---------|--------|--------|
| [OLMo 7B](https://github.com/allenai/OLMo-core/blob/main/src/scripts/official/OLMo3/OLMo-3-1025-7B-pretrain-1.py#L42) | 0.0003 | 4M |
| [DeepSeek 7B](https://arxiv.org/pdf/2401.02954) | 0.00042 | 9.44M|
| [Qwen 7B](https://arxiv.org/pdf/2309.16609) | 0.0003 | 4M |
| English |     0.00035    |  8M|
| Multilingual |    0.0001***    | 8M |

*** WIP
